API: Allow newaxis indexing for `array_api` arrays

honno opened the pull request:
Resolves #21373. ~I'm guessing CI will throw linting errors due to the already long lines that the diff touches, not quite sure what the situation is in this regard for `numpy.array_api`.~ PR ended up passing due to the rework.

Notably this PR also reworks `_validate_index()`, as I think this will be handy for maintenance. It was a good opportunity to rework error messages, namely to use preciser language in regards to Array API scope, and to highlight variables used. I would love to hear what @asmeurer and folks think about it.

~Running `array-api-tests` it looks like I missed something with 0 shape behaviour~ Problem was using `array_api` arrays for internal indexing with `np.ndarray` arrays, which doesn't work as expected for some edge cases. Fixed!

honno committed:
TST: Add test checking if newaxis indexing works for `array_api`

Also removes previous check against newaxis indexing, which is now outdated

honno committed:
TST, BUG: Allow `None` in `array_api` indexing

Introduces test for validating flat indexing when `None` is present

honno committed:
MAINT,DOC,TST: Rework of `_validate_index()` in `numpy.array_api`

_validate_index() is now called as self._validate_index(shape), and does not
return a key. This rework removes the recursive pattern used. Tests are
introduced to cover some edge cases. Additionally, its internal docstring
reflects new behaviour, and extends the flat indexing note.

asmeurer commented on numpy/array_api/_array_object.py:
```diff
@@ -292,100 +290,122 @@ def _validate_index(key, shape):
-                    "Multidimensional arrays must include an index for every axis or use an ellipsis"
+                    f"{key=} contains {n_single_axes} single-axis indices, "
+                    f"but should contain {self.ndim=} single-axis indices - "
+                    "an array key must index every axis."
```
```suggestion
                    "an array index must index every axis."
```

I wouldn't use the word "key" to refer to an index.

honno replied:
Yeah it's not ideal, I think I was avoiding saying index twice in two different ways. How about
```suggestion
                    "an array index must cover every axis."
```

asmeurer replied:
That's probably fine, although I think I prefer my original wording.

seberg replied:
I like mentioning ellipsis, because it offers the correct solution.  Maybe something like:
> The multidimensional index only specifies N dimensions, but the array has K dimensions.  If this is intentional, add an Ellipsis (`...`) which expands into as many slices (`:`) as necessary.

Saying how many dimensions were not indexed seems cute, but before we bike-shed a lot, I also think the original was just fine.

honno replied:
Thanks yeah that makes sense @seberg, pushed an inspired change if you and @asmeurer would like to take a look.

asmeurer commented on numpy/array_api/_array_object.py:
```diff
@@ -292,100 +290,122 @@ def _validate_index(key, shape):
                 raise IndexError(
-                    "Multidimensional arrays must include an index for every axis or use an ellipsis"
+                    f"Single-axes index {i} is a tuple, but nested tuple "
+                    "indices are not specified in the Array API."
```
This is maybe a little misleading, since this doesn't make sense in NumPy either.

asmeurer replied:
I guess NumPy does sometimes allow a tuple to represent an integer array. But it's deprecated.

honno replied:
> I guess NumPy does sometimes allow a tuple to represent an integer array. But it's deprecated.

Ah I didn't realise this was deprecated. I think it's better to be safe here, especially considering no warning is given when you use nested tuples for integer indexing e.g. `np.arange(5)[((0, 1),)]`.

asmeurer replied:
That's actually the reason why we handle it here at all. If NumPy errored, we could just pass it through to ndarray.

seberg replied:
The deprecation period is over in NumPy.  NumPy parses any tuple as a multidimensional index (or tries to), and _within_ a multidimensional index will convert any sequence to an array/advanced index.

asmeurer replied:
Maybe I'm misremembering what is deprecated. I would have assumed that the ideal behavior would be for a tuple inside of a tuple to be an error, and only treat a list as an array.

seberg replied:
tuples are special.  Once inside a tuple (n-dimensional index), NumPy allows any "array-like", and tuples are considered array-like as well after all.
That is legitimate to not allow all NumPy array-likes of course, but I think it makes sense for NumPy, even if it may be a bit surprising if you write `arr[(1, 2),]`.

seberg replied:
I half suspsect this whole branch is unreachable anyway with the "check tuple first" reorganization.

In any case, whether or not this could be refactored more, I think we should put this in once @asmeurer did a final review.

asmeurer commented on numpy/array_api/_array_object.py:
```diff
@@ -292,100 +290,122 @@ def _validate_index(key, shape):
+                elif i.dtype in _integer_dtypes and i.ndim != 0:
+                    raise IndexError(
+                        f"Single-axes index {i} is a non-zero-dimensional "
+                        "integer array, but advance integer indexing is not "
```
```suggestion
                        "integer array, but advanced integer indexing is not "
```

asmeurer commented on numpy/array_api/_array_object.py:
```diff
@@ -292,100 +290,122 @@ def _validate_index(key, shape):
+        n_single_axes = len(single_axes)
+        if n_ellipsis > 1:
+            raise IndexError(
+                f"{key=} contains {n_ellipsis} ellipses (...), but should "
```
Do we guard against the key having other stuff in it before this error message happens? If not, we should omit the `{key=}` part.

honno replied:
Oh, you mean like if say `key=[0, ..., xp.asarray(True), ..., 2.3]`, it'll get confusing? I think there's no way around that, seeing as we have to give an error at some point, and I think even then showing the key is useful to showcase how multiple ellipses have been used.

asmeurer replied:
Actually I just meant that if `key` could be anything, then you are potentially converting anything into a string for the error message. For instance, if `key` contains a very large array, then the error message could be huge. It's something you always have to be careful about when interpolating variables into error messages like this.

honno replied:
Oh yeah NumPy-proper already errors for this, so we should just fallback IMO. Pushed a change that does just that.

asmeurer commented:
I think the logic itself should be tested by the tests here, as well as the actual indexing tests in array_api_tests (you can check if anything is missing), so as long as those are passing I'm OK with the logic changes here.

honno committed:
MAINT: `advance` -> `advanced` (integer indexing)

Co-authored-by: Aaron Meurer <asmeurer@gmail.com>

honno committed:
BUG: array_api arrays use internal arrays from array_api array keys

When an array_api array is passed as the key for get/setitem, we access the
key's internal np.ndarray array to be used as the key for the internal
get/setitem operation. This behaviour was initially removed when
`_validate_index()` was reworked.

seberg commented on numpy/array_api/_array_object.py:
```diff
@@ -292,100 +290,122 @@ def _validate_index(key, shape):
+                f"{key=} contains {n_ellipsis} ellipses (...), but should "
+                "contain no more than 1."
+            )
+        elif n_ellipsis != 1:
```
```suggestion
        elif n_ellipsis == 0:
```
seems more clear to me spell "no ellipsis".

seberg reviewed (commented):
I have not dug too deep, but it seems LGTM modulo the error message discussion.  I would avoid the word "key" I think, because for me that means mappings (i.e. it only fits to structured arrays).
Ys, it is nice that it gives you a word for the full index, but my gut feeling is that it isn't ideal either.

The NumPy core code, I think works by walking through the dimension being indexed, as in `if index is ellipses: dim += N; elif index is None: continue; else: dim += 1`.  That might lead to simpler code, but I am not sure, so I will let you guys decide if you want to have a look.

honno committed:
MAINT: Better flat indexing error message for `array_api` arrays

Also better semantics for its prior ellipsis count condition

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

honno committed:
MAINT: `array_api` arrays don't special case multi-ellipsis errors

This gets handled by NumPy-proper.

asmeurer commented on numpy/array_api/tests/test_array_object.py:
```diff
@@ -322,3 +319,57 @@ def test___array__():
+
+def test_array_keys_use_private_array():
+    """
+    Indexing operations convert array keys before indexing the internal array
```
Let's not call indices "keys" here either. Maybe this is my fault for using "key" as the variable name in the code, which is just a habit for whenever I write a getitem or setitem, but it's probably one I should break for cases where the getitem is an index rather than a dict-like.

asmeurer commented:
The error messages look OK to me. Like I said, I'm not too worried about behavior because I think it should be fairly well tested between the tests here and the array API test suite indexing tests.

seberg commented on numpy/array_api/tests/test_array_object.py:
```diff
@@ -322,3 +319,57 @@ def test___array__():
+    in __getitem__(). This is achieved by passing array_api arrays with 0-sized
+    dimensions, which NumPy-proper treats erroneously - not sure why!
+
+    TODO: Find and use appropiate __setitem__() case.
```
This is a bug in NumPy proper!  Can you open an issue about it?  We probably might as well delete the test to be honest.
There is a simple 95% fix, the 100% fix shouldn't be particularly hard either though (but I have to look into it).

seberg replied:
OK, I opened gh-21462 about this.

seberg commented:
OK, lets just get this in for now.  The use of `key` in the test is a bit funny, but there are really worse things and maybe we will remember to clean it up the next time around ;).

Thanks Matthew and Aaron.
